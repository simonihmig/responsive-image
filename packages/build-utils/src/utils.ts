import baseN from 'base-n';
import sharp, { type Metadata } from 'sharp';
import type {
  ImageLoaderChainedResult,
  ImageOptions,
  LazyImageLoaderChainedResult,
} from './types';
import type { ImageConfig } from 'imagetools-core';

const b64 = baseN.create();

const defaultImageConfig: ImageOptions = {
  quality: 80,
  w: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
  // TODO: remove this, needs fixing tests
  allowUpscale: 'true',
  format: ['original', 'webp'],
};

const queryArraySeparator = ';';

export const META_KEY = 'responsive-image';

export function parseURL(id: string) {
  return new URL(id, 'file://');
}

export function getPathname(id: string) {
  const url = parseURL(id);
  return decodeURIComponent(url.pathname);
}

export function parseQuery(
  query: string | URLSearchParams,
): Partial<ImageOptions> {
  const params =
    query instanceof URLSearchParams ? query : new URLSearchParams(query);
  return Object.fromEntries(
    [...params.entries()].map(([key, value]) => {
      switch (key) {
        case 'w':
          return [
            key,
            value.split(queryArraySeparator).map((v) => parseInt(v, 10)),
          ];
        case 'format':
          return [key, value.split(queryArraySeparator)];
        case 'quality':
          return [key, parseInt(value, 10)];
        case 'lqip': {
          const parsedValue =
            value.charAt(0) === '{' ? JSON.parse(value) : value;

          return [
            key,
            typeof parsedValue === 'string'
              ? { type: parsedValue }
              : parsedValue,
          ];
        }

        default:
          return [key, value];
      }
    }),
  );
}

export async function getImagetoolsConfigs<
  BUILDOPTIONS extends Record<string, unknown>,
>(
  options: ImageOptions & BUILDOPTIONS,
  buildKeys: Array<keyof BUILDOPTIONS>,
): Promise<ImageConfig[]> {
  const imagetools = await import('imagetools-core');

  const entries = Object.entries(options)
    .filter(([key]) => !buildKeys.includes(key))
    .map(([key, value]) => {
      // imagetools expects this type
      const stringarrayifiedValue = Array.isArray(value)
        ? value.map((v) => String(v))
        : [String(value)];

      return [key, stringarrayifiedValue] satisfies [string, string[]];
    });

  return imagetools.resolveConfigs(entries, imagetools.builtinOutputFormats);
}

export function getOptions<BUILDOPTIONS>(
  id: string | URL,
  defaultOptions: BUILDOPTIONS,
  loaderOptions: Partial<BUILDOPTIONS>,
): ImageOptions & BUILDOPTIONS {
  const url = id instanceof URL ? id : parseURL(id);
  const parsedResourceQuery = parseQuery(url.searchParams);

  // Combines defaults, webpack options and query options,
  return {
    ...defaultImageConfig,
    ...defaultOptions,
    ...loaderOptions,
    ...parsedResourceQuery,
  } as ImageOptions & BUILDOPTIONS;
}

export function normalizeInput(
  input: string | Buffer | LazyImageLoaderChainedResult,
): LazyImageLoaderChainedResult;
export function normalizeInput(
  input: string | Buffer | ImageLoaderChainedResult,
): ImageLoaderChainedResult;
export function normalizeInput(
  input:
    | string
    | Buffer
    | ImageLoaderChainedResult
    | LazyImageLoaderChainedResult,
): ImageLoaderChainedResult | LazyImageLoaderChainedResult {
  if (typeof input === 'string') {
    return {
      images: [],
      sharp: sharp(getPathname(input)),
      imports: [],
    };
  }

  if (Buffer.isBuffer(input)) {
    return {
      images: [],
      sharp: sharp(input),
      imports: [],
    };
  }

  return input;
}

const generatedClassNames = new Map<string, string>();

export function generateLqipClassName(resource: string): string {
  if (generatedClassNames.has(resource)) {
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    return generatedClassNames.get(resource)!;
  } else {
    const className = `eri-dyn-${b64.encode(generatedClassNames.size)}`;
    generatedClassNames.set(resource, className);

    return className;
  }
}

export function getAspectRatio(meta: Metadata): number | undefined {
  if (meta.height && meta.width && meta.height > 0) {
    return Math.round((meta.width / meta.height) * 100) / 100;
  }
}

export function onlyUnique<T>(value: T, index: number, self: T[]): boolean {
  return self.indexOf(value) === index;
}

export function dataUri(
  data: string | Buffer,
  type: string,
  base64 = false,
): string {
  return `data:${type};base64,${
    base64 ? data : Buffer.from(data).toString('base64')
  }`;
}

export function blurrySvg(src: string, width: number, height: number): string {
  return `<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" viewBox="0 0 ${width} ${height}">
<filter id="b" color-interpolation-filters="sRGB"><feGaussianBlur stdDeviation=".5"></feGaussianBlur><feComponentTransfer><feFuncA type="discrete" tableValues="1 1"></feFuncA></feComponentTransfer></filter>
<image filter="url(#b)" preserveAspectRatio="none" height="100%" width="100%" xlink:href="${src}"></image>
</svg>`;
}
