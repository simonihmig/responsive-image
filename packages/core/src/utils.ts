import type { NonFunction, ValueOrCallback } from './types';

export function getValueOrCallback<T extends NonFunction>(
  valueOrCb: ValueOrCallback<T>,
): T {
  if (typeof valueOrCb === 'function') {
    return valueOrCb();
  } else {
    return valueOrCb;
  }
}
