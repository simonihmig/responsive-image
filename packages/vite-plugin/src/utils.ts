import {
  getOptions,
  type ImageLoaderChainedResult,
} from '@responsive-image/build-utils';

import type { Options, ViteOptions } from './types';
import type { PluginContext } from 'rollup';
import type { ResolvedConfig } from 'vite';

export const META_KEY = 'responsive-image';

export const defaultViteOptions = {
  name: '[name]-[width]w.[ext]',
  cache: true,
};

export const viteOptionKeys = [
  'exclude',
  'include',
  'lqip',
  'name',
] satisfies Array<keyof ViteOptions>;

export function getViteOptions(
  id: string | URL,
  loaderOptions: Partial<Options>,
): Options {
  return getOptions<ViteOptions>(id, defaultViteOptions, loaderOptions);
}

export function getInput(
  context: PluginContext,
  id: string,
): ImageLoaderChainedResult | undefined {
  const info = context.getModuleInfo(id);
  return info?.meta[META_KEY];
}

export function getViteBasePath(viteConfig: ResolvedConfig): string {
  return `${
    viteConfig.base.endsWith('/')
      ? viteConfig.base.slice(0, -1)
      : viteConfig.base
  }/@responsive-image/vite-plugin/`;
}
