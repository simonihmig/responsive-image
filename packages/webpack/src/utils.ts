import type { LoaderContext } from 'webpack';
import type { Options, WebpackLoaderOptions } from './types';
import {
  getOptions,
  type ImageLoaderChainedResult,
} from '@responsive-image/build-utils';

export const defaultWebpackOptions = {
  name: '[name]-[width]w-[hash].[ext]',
  outputPath: 'images',
};

export const webpackOptionKeys = [
  'lqip',
  'name',
  'outputPath',
  'webPath',
] satisfies Array<keyof WebpackLoaderOptions>;

export function getWebpackOptions(
  context: LoaderContext<Partial<Options>>,
): Options {
  return getOptions<WebpackLoaderOptions>(
    context.resource,
    defaultWebpackOptions,
    context.getOptions(),
  );
}

export function onlyUnique<T>(value: T, index: number, self: T[]): boolean {
  return self.indexOf(value) === index;
}

export function assertInput(
  input: string | Buffer | ImageLoaderChainedResult,
): asserts input is ImageLoaderChainedResult {
  if (Buffer.isBuffer(input) || typeof input === 'string') {
    throw new Error(
      'You cannot run this webpack loader on raw data, at least @responsive-image/loader is missing in the loader chain!',
    );
  }
}
