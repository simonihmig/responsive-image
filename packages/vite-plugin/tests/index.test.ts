import { describe, expect, test } from 'vitest';
import { configureToMatchImageSnapshot } from 'jest-image-snapshot';
import { compile } from './utils';

const customConfig = { threshold: 0.1 };
const toMatchImageSnapshot = configureToMatchImageSnapshot({
  customDiffConfig: customConfig,
});

expect.extend({ toMatchImageSnapshot });

describe('filter', () => {
  test('it operates on included assets', async () => {
    const { source } = await compile('image.jpg', {
      include: '**/*.jpg',
    });

    expect(source).toMatchSnapshot();
  });

  test('it skips non-matches', async () => {
    const { source, assets } = await compile('image.jpg', {
      include: '**/*.png',
    });

    expect(source).not.toContain('imageUrlFor');
    expect(assets).toHaveLength(0);
  });

  test('it skips excluded assets', async () => {
    const { source, assets } = await compile('image.jpg', {
      include: '**/*.jpg',
      exclude: '**/image.jpg',
    });

    expect(source).not.toContain('imageUrlFor');
    expect(assets).toHaveLength(0);
  });
});

test('it produces expected output', async () => {
  const { source, assets } = await compile('image.jpg', {
    format: ['png', 'webp'],
    include: '**/*.jpg',
  });

  expect(source).toMatchSnapshot();

  expect(assets.map((a) => a.fileName)).toEqual(
    [
      'image-640w.png',
      'image-640w.webp',
      'image-750w.png',
      'image-750w.webp',
      'image-828w.png',
      'image-828w.webp',
      'image-1080w.png',
      'image-1080w.webp',
      'image-1200w.png',
      'image-1200w.webp',
      'image-1920w.png',
      'image-1920w.webp',
      'image-2048w.png',
      'image-2048w.webp',
      'image-3840w.png',
      'image-3840w.webp',
    ].sort(),
  );

  expect(
    assets.find((a) => a.fileName === 'image-640w.png')?.source,
  ).toMatchImageSnapshot();
});

test('custom loader options are supported', async () => {
  const { source, assets } = await compile('image.jpg', {
    w: [100, 200],
    format: ['png'],
    name: 'test-[width].[ext]',
    include: '**/*.jpg',
  });

  expect(source).toMatchSnapshot();

  expect(assets.toSorted().map((a) => a.fileName)).toEqual([
    'test-100.png',
    'test-200.png',
  ]);

  for (const image of assets) {
    expect(image.source).toMatchImageSnapshot();
  }
});

test('custom query params are supported', async () => {
  const { source, assets } = await compile('image.jpg?w=100;200&format=png', {
    include: /^[^?]+\.jpg(\?.*)?$/,
  });

  expect(source).toMatchSnapshot();

  expect(assets.map((a) => a.fileName)).toEqual([
    'image-100w.png',
    'image-200w.png',
  ]);

  for (const image of assets) {
    expect(image.source).toMatchImageSnapshot();
  }
});

test('imagetools params are supported', async () => {
  const { source, assets } = await compile(
    'image.jpg?w=100;200&format=png&grayscale',
    {
      include: /^[^?]+\.jpg(\?.*)?$/,
    },
  );

  expect(source).toMatchSnapshot();

  expect(assets.map((a) => a.fileName)).toEqual([
    'image-100w.png',
    'image-200w.png',
  ]);

  for (const image of assets) {
    expect(image.source).toMatchImageSnapshot();
  }
});

describe('LQIP', async () => {
  test('color LQIP is supported', async () => {
    const { source, assets } = await compile('image.jpg', {
      include: '**/*.jpg',
      w: [100, 200],
      lqip: { type: 'color' },
    });

    expect(assets.map((a) => a.fileName)).toEqual([
      'image-100w.jpg',
      'image-100w.webp',
      'image-200w.jpg',
      'image-200w.webp',
      'style.css',
    ]);

    expect(source).toMatchSnapshot();

    const style = assets.find((a) => a.fileName === 'style.css');
    expect(style?.source).toMatchInlineSnapshot(`
      ".ri-dyn-0{background-color:#083838}
      "
    `);
  });

  test('inline LQIP is supported', async () => {
    const { source, assets } = await compile('image.jpg', {
      include: '**/*.jpg',
      w: [100, 200],
      lqip: { type: 'inline' },
    });

    expect(assets.map((a) => a.fileName)).toEqual([
      'image-100w.jpg',
      'image-100w.webp',
      'image-200w.jpg',
      'image-200w.webp',
      'style.css',
    ]);

    expect(source).toMatchSnapshot();

    const style = assets.find((a) => a.fileName === 'style.css');
    expect(style?.source).toMatchInlineSnapshot(`
      ".ri-dyn-0{background-image:url(data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHhtbG5zOnhsaW5rPSJodHRwOi8vd3d3LnczLm9yZy8xOTk5L3hsaW5rIiB2aWV3Qm94PSIwIDAgNjQwIDQyNyI+CjxmaWx0ZXIgaWQ9ImIiIGNvbG9yLWludGVycG9sYXRpb24tZmlsdGVycz0ic1JHQiI+PGZlR2F1c3NpYW5CbHVyIHN0ZERldmlhdGlvbj0iLjUiPjwvZmVHYXVzc2lhbkJsdXI+PGZlQ29tcG9uZW50VHJhbnNmZXI+PGZlRnVuY0EgdHlwZT0iZGlzY3JldGUiIHRhYmxlVmFsdWVzPSIxIDEiPjwvZmVGdW5jQT48L2ZlQ29tcG9uZW50VHJhbnNmZXI+PC9maWx0ZXI+CjxpbWFnZSBmaWx0ZXI9InVybCgjYikiIHByZXNlcnZlQXNwZWN0UmF0aW89Im5vbmUiIGhlaWdodD0iMTAwJSIgd2lkdGg9IjEwMCUiIHhsaW5rOmhyZWY9ImRhdGE6aW1hZ2UvcG5nO2Jhc2U2NCxpVkJPUncwS0dnb0FBQUFOU1VoRVVnQUFBQWtBQUFBR0NBSUFBQUNlcFNPU0FBQUFDWEJJV1hNQUFBc1RBQUFMRXdFQW1wd1lBQUFBc0VsRVFWUjRuR05nTlRmaDFERGlrdFBoa05WbTB6UmlkWE5pczdaaDFURm0xamRpWVBWMzR2Rnc0ZGUxNWxVeTQ5SzM1QTN6NDNCMzVYRnc0QS96WkdEemMrZkxEaGVKQ3hXMjhCQXdkcFBLU1pDTkRWTlBDeFh5c21mZ01MUGw4M1NVakhkVktReTF5QTVOMlpoclhSWW1iR2ZGcHFMT3dLYW54YVdudzZ1cEx1dGxaSlpwWVpGaEkrWmt3V2xveks2cHpjQWlMODhxTDgydXA4Qm5yeVh2cVduWVlDQVpyTWR0WnNwbWFnSUFBaFFmUDJab1lSQUFBQUFBU1VWT1JLNUNZSUk9Ij48L2ltYWdlPgo8L3N2Zz4=)}
      "
    `);
  });

  test('blurhash LQIP is supported', async () => {
    const { source } = await compile('image.jpg', {
      include: '**/*.jpg',
      w: [100, 200],
      lqip: { type: 'blurhash' },
    });

    expect(source).toMatchSnapshot();
  });
});
