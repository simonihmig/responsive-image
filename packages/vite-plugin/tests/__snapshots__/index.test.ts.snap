// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LQIP > blurhash LQIP is supported 1`] = `
"function r(e, n, a) {
  const o = e.map((t) => t.width).reduce((t, i) => i >= n && t >= n ? i >= t ? t : i : i >= t ? i : t, 0);
  return e.find(
    (t) => t.width === o && t.format === a
  );
}
const h = /* @__PURE__ */ new Map();
function m(e) {
  return h.get(e);
}
var g;
(g = m("env")) == null || g.deviceWidths;
const p = [{ url: "/image-100w.jpg", width: 100, format: "jpeg" }, { url: "/image-100w.webp", width: 100, format: "webp" }, { url: "/image-200w.jpg", width: 200, format: "jpeg" }, { url: "/image-200w.webp", width: 200, format: "webp" }], c = {
  imageTypes: ["jpeg", "webp"],
  availableWidths: [100, 200],
  lqip: { type: "blurhash", hash: "M53dVIR8D8*Ft2MpoxylRoRlHYniyBRQbu", width: 5, height: 3 },
  aspectRatio: 1.5,
  imageUrlFor(e, n) {
    var a;
    return (a = r(p, e, n ?? "jpeg")) == null ? void 0 : a.url;
  }
};
console.log(c);
"
`;

exports[`LQIP > color LQIP is supported 1`] = `
"function o(t, n, a) {
  const r = t.map((e) => e.width).reduce((e, i) => i >= n && e >= n ? i >= e ? e : i : i >= e ? i : e, 0);
  return t.find(
    (e) => e.width === r && e.format === a
  );
}
const c = /* @__PURE__ */ new Map();
function m(t) {
  return c.get(t);
}
var g;
(g = m("env")) == null || g.deviceWidths;
const p = [{ url: "/image-100w.jpg", width: 100, format: "jpeg" }, { url: "/image-100w.webp", width: 100, format: "webp" }, { url: "/image-200w.jpg", width: 200, format: "jpeg" }, { url: "/image-200w.webp", width: 200, format: "webp" }], s = {
  imageTypes: ["jpeg", "webp"],
  availableWidths: [100, 200],
  lqip: { type: "color", class: "ri-dyn-0" },
  aspectRatio: 1.5,
  imageUrlFor(t, n) {
    var a;
    return (a = o(p, t, n ?? "jpeg")) == null ? void 0 : a.url;
  }
};
console.log(s);
"
`;

exports[`LQIP > inline LQIP is supported 1`] = `
"function o(t, n, a) {
  const r = t.map((e) => e.width).reduce((e, i) => i >= n && e >= n ? i >= e ? e : i : i >= e ? i : e, 0);
  return t.find(
    (e) => e.width === r && e.format === a
  );
}
const m = /* @__PURE__ */ new Map();
function c(t) {
  return m.get(t);
}
var g;
(g = c("env")) == null || g.deviceWidths;
const p = [{ url: "/image-100w.jpg", width: 100, format: "jpeg" }, { url: "/image-100w.webp", width: 100, format: "webp" }, { url: "/image-200w.jpg", width: 200, format: "jpeg" }, { url: "/image-200w.webp", width: 200, format: "webp" }], s = {
  imageTypes: ["jpeg", "webp"],
  availableWidths: [100, 200],
  lqip: { type: "inline", class: "ri-dyn-0" },
  aspectRatio: 1.5,
  imageUrlFor(t, n) {
    var a;
    return (a = o(p, t, n ?? "jpeg")) == null ? void 0 : a.url;
  }
};
console.log(s);
"
`;

exports[`custom loader options are supported 1`] = `
"function a(n, e, g) {
  const r = n.map((t) => t.width).reduce((t, i) => i >= e && t >= e ? i >= t ? t : i : i >= t ? i : t, 0);
  return n.find(
    (t) => t.width === r && t.format === g
  );
}
const c = /* @__PURE__ */ new Map();
function s(n) {
  return c.get(n);
}
var o;
(o = s("env")) == null || o.deviceWidths;
const d = [{ url: "/test-100.png", width: 100, format: "png" }, { url: "/test-200.png", width: 200, format: "png" }], f = {
  imageTypes: ["png"],
  availableWidths: [100, 200],
  aspectRatio: 1.5,
  imageUrlFor(n, e) {
    var g;
    return (g = a(d, n, e ?? "png")) == null ? void 0 : g.url;
  }
};
console.log(f);
"
`;

exports[`custom query params are supported 1`] = `
"function r(t, e, g) {
  const o = t.map((n) => n.width).reduce((n, i) => i >= e && n >= e ? i >= n ? n : i : i >= n ? i : n, 0);
  return t.find(
    (n) => n.width === o && n.format === g
  );
}
const c = /* @__PURE__ */ new Map();
function m(t) {
  return c.get(t);
}
var a;
(a = m("env")) == null || a.deviceWidths;
const s = [{ url: "/image-100w.png", width: 100, format: "png" }, { url: "/image-200w.png", width: 200, format: "png" }], d = {
  imageTypes: ["png"],
  availableWidths: [100, 200],
  aspectRatio: 1.5,
  imageUrlFor(t, e) {
    var g;
    return (g = r(s, t, e ?? "png")) == null ? void 0 : g.url;
  }
};
console.log(d);
"
`;

exports[`filter > it operates on included assets 1`] = `
"function m(t, w, g) {
  const r = t.map((e) => e.width).reduce((e, i) => i >= w && e >= w ? i >= e ? e : i : i >= e ? i : e, 0);
  return t.find(
    (e) => e.width === r && e.format === g
  );
}
const p = /* @__PURE__ */ new Map();
function o(t) {
  return p.get(t);
}
var a;
(a = o("env")) == null || a.deviceWidths;
const d = [{ url: "/image-640w.jpg", width: 640, format: "jpeg" }, { url: "/image-640w.webp", width: 640, format: "webp" }, { url: "/image-750w.jpg", width: 750, format: "jpeg" }, { url: "/image-750w.webp", width: 750, format: "webp" }, { url: "/image-828w.jpg", width: 828, format: "jpeg" }, { url: "/image-828w.webp", width: 828, format: "webp" }, { url: "/image-1080w.jpg", width: 1080, format: "jpeg" }, { url: "/image-1080w.webp", width: 1080, format: "webp" }, { url: "/image-1200w.jpg", width: 1200, format: "jpeg" }, { url: "/image-1200w.webp", width: 1200, format: "webp" }, { url: "/image-1920w.jpg", width: 1920, format: "jpeg" }, { url: "/image-1920w.webp", width: 1920, format: "webp" }, { url: "/image-2048w.jpg", width: 2048, format: "jpeg" }, { url: "/image-2048w.webp", width: 2048, format: "webp" }, { url: "/image-3840w.jpg", width: 3840, format: "jpeg" }, { url: "/image-3840w.webp", width: 3840, format: "webp" }], f = {
  imageTypes: ["jpeg", "webp"],
  availableWidths: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
  aspectRatio: 1.5,
  imageUrlFor(t, w) {
    var g;
    return (g = m(d, t, w ?? "jpeg")) == null ? void 0 : g.url;
  }
};
console.log(f);
"
`;

exports[`imagetools params are supported 1`] = `
"function r(t, e, g) {
  const o = t.map((n) => n.width).reduce((n, i) => i >= e && n >= e ? i >= n ? n : i : i >= n ? i : n, 0);
  return t.find(
    (n) => n.width === o && n.format === g
  );
}
const c = /* @__PURE__ */ new Map();
function m(t) {
  return c.get(t);
}
var a;
(a = m("env")) == null || a.deviceWidths;
const s = [{ url: "/image-100w.png", width: 100, format: "png" }, { url: "/image-200w.png", width: 200, format: "png" }], d = {
  imageTypes: ["png"],
  availableWidths: [100, 200],
  aspectRatio: 1.5,
  imageUrlFor(t, e) {
    var g;
    return (g = r(s, t, e ?? "png")) == null ? void 0 : g.url;
  }
};
console.log(d);
"
`;

exports[`it produces expected output 1`] = `
"function m(e, w, g) {
  const r = e.map((t) => t.width).reduce((t, i) => i >= w && t >= w ? i >= t ? t : i : i >= t ? i : t, 0);
  return e.find(
    (t) => t.width === r && t.format === g
  );
}
const n = /* @__PURE__ */ new Map();
function p(e) {
  return n.get(e);
}
var a;
(a = p("env")) == null || a.deviceWidths;
const o = [{ url: "/image-640w.png", width: 640, format: "png" }, { url: "/image-640w.webp", width: 640, format: "webp" }, { url: "/image-750w.png", width: 750, format: "png" }, { url: "/image-750w.webp", width: 750, format: "webp" }, { url: "/image-828w.png", width: 828, format: "png" }, { url: "/image-828w.webp", width: 828, format: "webp" }, { url: "/image-1080w.png", width: 1080, format: "png" }, { url: "/image-1080w.webp", width: 1080, format: "webp" }, { url: "/image-1200w.png", width: 1200, format: "png" }, { url: "/image-1200w.webp", width: 1200, format: "webp" }, { url: "/image-1920w.png", width: 1920, format: "png" }, { url: "/image-1920w.webp", width: 1920, format: "webp" }, { url: "/image-2048w.png", width: 2048, format: "png" }, { url: "/image-2048w.webp", width: 2048, format: "webp" }, { url: "/image-3840w.png", width: 3840, format: "png" }, { url: "/image-3840w.webp", width: 3840, format: "webp" }], d = {
  imageTypes: ["png", "webp"],
  availableWidths: [640, 750, 828, 1080, 1200, 1920, 2048, 3840],
  aspectRatio: 1.5,
  imageUrlFor(e, w) {
    var g;
    return (g = m(o, e, w ?? "png")) == null ? void 0 : g.url;
  }
};
console.log(d);
"
`;
